core-bundle/src/Menu/FrontendMenuBuilder.php

```diff
@@ -0,0 +1,371 @@
+
+    /**
+     * @return array<array{page:PageModel,hasSubpages:bool}>
+     */
+    private function findPagesByIds(array $pageIds): array
+    {
+        // Get all active pages and also include root pages if the language is added to the URL (see #72)
+        $pages = $this->pageModelAdapter->findPublishedRegularByIds($pageIds, ['includeRoot' => true]);
+
+        if (null === $pages) {
+            return [];
+        }
+
+        return array_map(
+            static fn (PageModel $page): array => [
+                'page' => $page,
+                'hasSubpages' => false,
+            ],
+            $pages instanceof \Traversable ? iterator_to_array($pages) : $pages
+        );
+    }
+
+    /**
+     * @return array<array{page:PageModel,hasSubpages:bool}>
+     */
+    private function findPagesByPid(int $pid, bool $showHidden = false, bool $isSitemap = false): array
+    {
+        $time = Date::floorToMinute();
+        $blnBeUserLoggedIn = $this->tokenChecker->hasBackendUser() && $this->tokenChecker->isPreviewMode();
+        $unroutableTypes = $this->pageRegistry->getUnroutableTypes();
+
+        $pages = $this->connection
+            ->executeQuery("SELECT p1.id, EXISTS(SELECT * FROM tl_page p2 WHERE p2.pid=p1.id AND p2.type!='root' AND p2.type NOT IN ('".implode("', '", $unroutableTypes)."')".(!$showHidden ? ($isSitemap ? " AND (p2.hide='' OR sitemap='map_always')" : " AND p2.hide=''") : '').(!$blnBeUserLoggedIn ? " AND p2.published='1' AND (p2.start='' OR p2.start<='$time') AND (p2.stop='' OR p2.stop>'$time')" : '').") AS hasSubpages FROM tl_page p1 WHERE p1.pid=:pid AND p1.type!='root' AND p1.type NOT IN ('".implode("', '", $unroutableTypes)."')".(!$showHidden ? ($isSitemap ? " AND (p1.hide='' OR sitemap='map_always')" : " AND p1.hide=''") : '').(!$blnBeUserLoggedIn ? " AND p1.published='1' AND (p1.start='' OR p1.start<='$time') AND (p1.stop='' OR p1.stop>'$time')" : '').' ORDER BY p1.sorting', ['pid' => $pid])
+            ->fetchAllAssociative()
+        ;
+
+        if (\count($pages) < 1) {
+            return [];
+        }
+
+        // Load models into the registry with a single query
+        $this->pageModelAdapter->findMultipleByIds(array_map(static fn ($row) => $row['id'], $pages));
+
+        return array_map(fn (array $row): array => [
+            'page' => $this->pageModelAdapter->findByPk($row['id']),
+            'hasSubpages' => (bool) $row['hasSubpages'],
+        ], $pages);
+    }
+
+    private function generateUri(PageModel $pageModel, ItemInterface $menuItem): ?string
+    {
+        if ('redirect' === $pageModel->type) {
+            $href = $pageModel->url;
+
+            if (0 === strncasecmp($href, 'mailto:', 7)) {
+                return StringUtil::encodeEmail($href);
+            }
+
+            return $href;
+        }
+
+        if ('root' === $pageModel->type) {
+            // Overwrite the alias to link to the empty URL or language URL (see #1641)
+            $pageModel->alias = 'index';
```

Comment:
This is the reference: https://github.com/contao/contao/blob/846437e41504ce7b0c864fe73436d68f6f5aaaf6/core-bundle/src/Resources/contao/modules/ModuleCustomnav.php#L113
but maybe this is some legacy stuff as well.

Reply:
Yeah, it's to make the root URL pretty, in case the page is rendered as part of a navigation. Otherwise the URL will be `https://example.com/alias-of-the-root-page` instead of just `https://example.com`.

Reply:
I understand what it does, but its still wrong. We would need to fetch the first published subpage and render that instead. Apart from the fact that re-implementing this all the time is a PITA and should be done by routing, but unfortunately that does not work (yet) 🙄 (see https://github.com/contao/contao/pull/2031)